src/txn/txn_rollback_to_stable.c

```diff
@@ -0,0 +1,396 @@
+			__txn_abort_newer_row_skip(
+			    session, insert, rollback_timestamp);
+	}
+}
+
+/*
+ * __txn_abort_newer_updates --
+ *	Abort updates on this page newer than the timestamp.
+ */
+static int
+__txn_abort_newer_updates(
+    WT_SESSION_IMPL *session, WT_REF *ref, wt_timestamp_t *rollback_timestamp)
+{
+	WT_PAGE *page;
+
+	page = ref->page;
+	switch (page->type) {
+	case WT_PAGE_ROW_INT:
+		/*
+		 * There is nothing to do for internal pages, since we aren't
+		 * rolling back far enough to potentially include reconciled
+		 * changes - and thus won't need to roll back structure
+		 * changes on internal pages.
+		 */
+		break;
+	case WT_PAGE_ROW_LEAF:
+		__txn_abort_newer_row_leaf(session, page, rollback_timestamp);
+		break;
+	default:
+		WT_RET_MSG(session, EINVAL, "rollback_to_stable "
+		    "is only supported for row store btrees");
+	}
+
+	return (0);
+}
+
+/*
+ * __txn_rollback_to_stable_custom_skip --
+ *	Return if custom rollback requires we read this page.
+ */
+static int
+__txn_rollback_to_stable_custom_skip(
+    WT_SESSION_IMPL *session, WT_REF *ref, void *context, bool *skipp)
+{
+	WT_UNUSED(session);
+	WT_UNUSED(context);
+
+	/* Review all pages that are in memory. */
+	if (ref->state == WT_REF_MEM || ref->state == WT_REF_DELETED)
+		*skipp = false;
+	else
+		*skipp = true;
+	return (0);
+}
+
+/*
+ * __txn_rollback_to_stable_btree_walk --
+ *	Called for each open handle - choose to either skip or wipe the commits
+ */
+static int
+__txn_rollback_to_stable_btree_walk(
+    WT_SESSION_IMPL *session, wt_timestamp_t *rollback_timestamp)
+{
+	WT_DECL_RET;
+	WT_PAGE *page;
+	WT_REF *ref;
+
+	/* Walk the tree, marking commits aborted where appropriate. */
+	ref = NULL;
+	while ((ret = __wt_tree_walk_custom_skip(session, &ref,
+	    __txn_rollback_to_stable_custom_skip,
+	    NULL, WT_READ_NO_EVICT)) == 0 && ref != NULL) {
+		page = ref->page;
+
+		/* Review deleted page saved to the ref */
+		if (ref->page_del != NULL && __wt_timestamp_cmp(
+		    rollback_timestamp, &ref->page_del->timestamp) < 0)
+			__wt_delete_page_rollback(session, ref);
+
+		if (!__wt_page_is_modified(page))
+			continue;
+
+		WT_RET(__txn_abort_newer_updates(
+		    session, ref, rollback_timestamp));
+	}
+	return (ret);
+}
+
+/*
+ * __txn_rollback_to_stable_btree --
+ *	Called for each open handle - choose to either skip or wipe the commits
+ */
+static int
+__txn_rollback_to_stable_btree(
+    WT_SESSION_IMPL *session, const char *cfg[])
+{
+	WT_CONFIG_ITEM cval;
+	WT_DECL_RET;
+	WT_DECL_TIMESTAMP(rollback_timestamp)
+	WT_BTREE *btree;
+
+	btree = S2BT(session);
+
+	/* Logged files don't get their commits wiped - that wouldn't be safe */
+	if (FLD_ISSET(S2C(session)->log_flags, WT_CONN_LOG_ENABLED) &&
+	    !F_ISSET(btree, WT_BTREE_NO_LOGGING)) {
+		/*
+		 * Add the btree ID to the bitstring, so we can exclude any
+		 * lookaside entries for this btree.
+		 */
+		__bit_set(
+		    S2C(session)->stable_rollback_bitstring, btree->id);
+		return (0);
```

Comment:
Mostly to check my understanding, this is the case where timestamps can continue to exist in the system that are greater than `rollback_timestamp`.  In particular, MongoDB's oplog will have entries more recent than the rollback point, and they continue to have timestamps in the future of the rollback point.

We rely on replication sorting that out by truncating the oplog.  We're then relying on the truncate being seen by all future readers to avoid parts of the pre-rollback oplog being resurrected.  As discussed, there are a lot of moving parts involved in getting that right, but I don't see any big simplifications given the repl / storage interface.